src/AccountManager.mo

```diff
@@ -491,40 +497,95 @@ module {
       };
     };
 
-    /// Initiates a withdrawal by transferring tokens to another account.
-    /// Returns ICRC1 transaction index and amount of transferred tokens (fee excluded).
-    public func withdraw(to : ICRC1.Account, amount : Nat) : async* WithdrawResponse {
-
-      totalWithdrawn_ += amount;
-
-      let callResult = await* processWithdrawTransfer(to, amount);
+    func handleWithdrawal(
+      from_subaccount : ?Blob,
+      to : ICRC1.Account,
+      amount : Nat,
+      onSuccess : (txIdx : Nat) -> async* WithdrawResponse,
+      onError : (err : ProcessWithdrawTransferError) -> async* WithdrawResponse,
+    ) : async* WithdrawResponse {
+      let callResult = await* processWithdrawTransfer(from_subaccount, to, amount);
 
       switch (callResult) {
-        case (#Ok txIdx) {
-          log(ownPrincipal, #withdraw({ to = to; amount = amount }));
-          #ok(txIdx, amount - fee(#withdrawal));
-        };
+        case (#Ok txIdx) { await* onSuccess(txIdx) };
         case (#Err(#BadFee { expected_fee })) {
           updateFee(expected_fee);
-          let retryResult = await* processWithdrawTransfer(to, amount);
+          let retryResult = await* processWithdrawTransfer(from_subaccount, to, amount);
           switch (retryResult) {
-            case (#Ok txIdx) {
+            case (#Ok txIdx) { await* onSuccess(txIdx) };
+            case (#Err err) { await* onError(err) };
+          };
+        };
+        case (#Err err) { await* onError(err) };
+      };
+    };
+
+    /// Initiates a direct withdrawal by transferring tokens to another account.
+    /// The transfer is made directly from the main account to the specified account indirectly.
+    /// In order to make the transfer traceable, the deposit account is used as an intermediary (fee x2).
+    /// Returns ICRC1 transaction index and amount of transferred tokens (fee excluded).
+    public func withdrawTraceably(p : Principal, to : ICRC1.Account, amount : Nat) : async* WithdrawResponse {
+      let ?release = depositRegistry.obtainLock(p) else return #err(#NotAvailable);
+
+      let depositAccount = {
+        owner = ownPrincipal;
+        subaccount = ?Util.toSubaccount(p);
+      };
+
+      let deposit = depositRegistry.get(p);
```

Comment:
I think we should inspect this value and use it. If the `deposit` value is larger than the withdrawal then we can skip the first transfer into the deposit account.

If the `deposit` is <= the withdrawal amount then we can "top up" the deposit account with precisely the difference and we can cancel the consolidation.

Important note: This is not merely an optimization. I think it will simplify the retry logic. Say the second step of the transfer failed (for example by fee change) and left an amount in the deposit account. When we retry everything from the beginning then the logic will pick up the amount left from the first attempt, top up the difference and withdraw. Otherwise we would either consolidate the amount first or temporarily make a double transfer into the deposit account.

Final note: there is a fundamental problem which is solved by utilizing the `deposit` value. If there is a only a single user and his amount has not been consolidated yet then we cannot process a withdrawal because there aren't any funds in the pool. So we have to use the `deposit` value.